python/mxnet/gluon/contrib/estimator/event_handler.py

```diff
@@ -0,0 +1,459 @@
+        self.batch_index = 0
+        self.current_epoch = 0
+        self.processed_samples = 0
+        # logging handler need to be called at last to make sure all states are updated
+        # it will also shut down logging at train end
+        self.priority = np.Inf
+
+    def train_begin(self, estimator, *args, **kwargs):
+        self.train_start = time.time()
+        trainer = estimator.trainer
+        optimizer = trainer.optimizer.__class__.__name__
+        lr = trainer.learning_rate
+        self.logger.info("Training begin: using optimizer %s "
+                         "with current learning rate %.4f ",
+                         optimizer, lr)
+        self.logger.info("Train for %d epochs.", estimator.max_epochs)
+
+    def train_end(self, estimator, *args, **kwargs):
+        train_time = time.time() - self.train_start
+        msg = 'Train finished using total %ds with %d epochs.' % (train_time, self.current_epoch)
+        # log every result in train stats including train/validation loss & metrics
+        for metric in self.train_metrics + self.val_metrics:
+            name, value = metric.get()
+            msg += '%s : %.4f ' % (name, value)
+        self.logger.info(msg)
+        for handler in self.logger.handlers:
+            handler.close()
+            self.logger.removeHandler(handler)
+        logging.shutdown()
+
+    def batch_begin(self, estimator, *args, **kwargs):
+        if self.verbose == self.LOG_VERBOSITY_PER_BATCH:
+            self.batch_start = time.time()
+
+    def batch_end(self, estimator, *args, **kwargs):
+        if self.verbose == self.LOG_VERBOSITY_PER_BATCH:
+            batch_time = time.time() - self.batch_start
+            msg = '[Epoch %d] [Batch %d] ' % (self.current_epoch, self.batch_index)
+            self.processed_samples += kwargs['batch'][0].shape[0]
+            msg += '[Samples %s] ' % (self.processed_samples)
+            msg += 'time/batch: %.3fs ' % batch_time
+            for metric in self.train_metrics:
+                # only log current training loss & metric after each batch
+                name, value = metric.get()
+                msg += '%s : %.4f ' % (name, value)
+            self.logger.info(msg)
+            self.batch_index += 1
+
+    def epoch_begin(self, estimator, *args, **kwargs):
+        if self.verbose >= self.LOG_VERBOSITY_PER_EPOCH:
+            self.epoch_start = time.time()
+
+    def epoch_end(self, estimator, *args, **kwargs):
+        if self.verbose >= self.LOG_VERBOSITY_PER_EPOCH:
+            epoch_time = time.time() - self.epoch_start
+            msg = '\n[Epoch %d] finished in %.3fs: ' % (self.current_epoch, epoch_time)
+            for monitor in self.train_metrics + self.val_metrics:
+                name, value = monitor.get()
+                msg += '%s : %.4f ' % (name, value)
+            self.logger.info(msg)
+            self.current_epoch += 1
+            self.batch_index = 0
+
+
+class CheckpointHandler(BatchEnd, EpochEnd):
```

Comment:
I don't see how an estimator can use the produced checkpoints and resume training from there. Is there any intention to cover that usage? With the tensorflow estimator api https://www.tensorflow.org/guide/checkpoints, it's very easy - as long as the `model_dir` argument is set. User don't need to change any code. 
To do that we also need to add logic to save trainer state and resume model parameter/trainer states.

Reply:
as we discussed offline before, we need to decide to save a few more information (serialize train history) in addition to params and trainer state. WIll need a bit of design, tacked https://issues.apache.org/jira/browse/MXNET-1373, will open another PR to add this feature after we can finalize general event handler design.

Reply:
added functionality to save symbol, param, and trainer states in model_dir, only keep a maximum of check points (default 5).  resuming training can be done mannually now, need to design for automatical resuming later.